### PR TITLE
[SPARK-44813][INFRA] The Jira Python misses our assignee when it searches users again

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR creates an alternative to the assign_issue function in jira.client.JIRA.
 
The original one has an issue that it will search users again and only choose the assignee from 20 candidates. If it's unmatched, it picks the head blindly.

For example,

```python
>>> assignee = asf_jira.user("yao")
>>> "SPARK-44801"
'SPARK-44801'
>>> asf_jira.assign_issue(issue.key, assignee.name)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'issue' is not defined
>>> asf_jira.assign_issue("SPARK-44801", assignee.name)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hzyaoqin/python/lib/python3.11/site-packages/jira/client.py", line 123, in wrapper
    result = func(*arg_list, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hzyaoqin/python/lib/python3.11/site-packages/jira/client.py", line 1891, in assign_issue
    self._session.put(url, data=json.dumps(payload))
  File "/Users/hzyaoqin/python/lib/python3.11/site-packages/requests/sessions.py", line 649, in put
    return self.request("PUT", url, data=data, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hzyaoqin/python/lib/python3.11/site-packages/jira/resilientsession.py", line 246, in request
    elif raise_on_error(response, **processed_kwargs):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hzyaoqin/python/lib/python3.11/site-packages/jira/resilientsession.py", line 71, in raise_on_error
    raise JIRAError(
jira.exceptions.JIRAError: JiraError HTTP 400 url: https://issues.apache.org/jira/rest/api/latest/issue/SPARK-44801/assignee
	response text = {"errorMessages":[],"errors":{"assignee":"User 'airhot' cannot be assigned issues."}}
```

The Jira userid 'yao' fails to return my JIRA profile as a candidate(20 in total) to match. So, 'airhot' from the head replaces me as an assignee.

### Why are the changes needed?

bugfix for merge_spark_pr

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

test locally

```python
>>> def assign_issue(client: jira.client.JIRA, issue: int, assignee: str) -> bool:
...     """Assign an issue to a user.
...
...     Args:
...         issue (Union[int, str]): the issue ID or key to assign
...         assignee (str): the user to assign the issue to. None will set it to unassigned. -1 will set it to Automatic.
...
...     Returns:
...         bool
...     """
...     url = getattr(client, "_get_latest_url")(f"issue/{issue}/assignee")
...     payload = {"name": assignee}
...     getattr(client, "_session").put(url, data=json.dumps(payload))
...     return True
...

>>>
>>> assign_issue(asf_jira, "SPARK-44801", "yao")
True
```

### CODE REVIEW

**Comment by yaooqinn** [comment]
FYI https://github.com/pycontribs/jira/blob/main/jira/client.py#L1862

cc @HyukjinKwon @cloud-fan @dongjoon-hyun, thanks

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @yaooqinn . Although this is orthogonal, shall we hold on any further change until we recover the merge script? There is a missing part in the following.
- https://github.com/apache/spark/pull/42484

I made a follow-up here.
- https://github.com/apache/spark/pull/42502

**Comment by srowen** [comment]
Let's re-run tests just to be sure, but i think we can proceed now and I can't see that the test failure can be related

**Comment by dongjoon-hyun** [comment]
Ya, we can proceed now.

**Review by cloud-fan — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master/3.5/3.4/3.3
@yaooqinn when I merged this, I was using the new changes.
I got an error after selecting your name, `{"assignee":"User 'airhot' cannot be assigned issues."}}`. That isn't your username :)  I wonder if there is still an issue with the new logic? but I assume this is a step forward

**Comment by yaooqinn** [comment]
```log
Would you like to update an associated JIRA? (y/n): y
Enter a JIRA id [SPARK-44813]:
JIRA is unassigned, choose assignee
[0] Kent Yao (Reporter)
[1] Sean R. Owen (Commentator)
Enter number of user, or userid, to assign to (blank to leave unassigned):0
=== JIRA SPARK-44813 ===
summary		The JIRA Python misses our assignee when it searches user again
assignee	Kent Yao
status		Reopened
url		https://issues.apache.org/jira/browse/SPARK-44813

Enter comma-separated fix version(s) [4.0.0,3.5.0,3.4.2,3.3.3]:
Successfully resolved SPARK-44813 with fixVersions=['4.0.0', '3.5.0', '3.4.2', '3.3.3']!
```

This is reverted, fixed on the caller side, and merged to master/3.5/3.4/3.3 again.

@srowen, the PR had an issue when you were merging. But, according to your error, you used the old merging script.

With the last update, it's OK now.

**Comment by dongjoon-hyun** [comment]
BTW, may I ask why this is backported to old branches? This script is supposed to be used in the `master` branch only, isn't it?

**Comment by srowen** [comment]
That's probably my mistake. I thought the old version might be used when you work on an old branch to back-port, but... wait maybe that isn't true at all. You would run the master version right? oops. Probably no harm, just wasn't needed

**Comment by dongjoon-hyun** [comment]
Ya, there is no harm, of course.

File: dev/merge_spark_pr.py [hunk 1]
```diff
@@ -403,6 +403,19 @@ def choose_jira_assignee(issue, asf_jira):
             print("Error assigning JIRA, try again (or leave blank and fix manually)")
 
 
+def assign_issue(client: jira.client.JIRA, issue: int, assignee: str) -> bool:
```
Comment (ueshin) [inline dev/merge_spark_pr.py:406]: Hi, this breaks the environment where `jira` is not installed.

Comment (cloud-fan) [reply dev/merge_spark_pr.py:406]: hmm, AFAIK the merge script required `jira` before this PR, as we need to update the JIRA ticket status.

Comment (ueshin) [reply dev/merge_spark_pr.py:406]: It still worked, just skipped to update the JIRA ticket status.
It's fine if it's a required dependency, though.

Comment (yaooqinn) [reply dev/merge_spark_pr.py:406]: I apologize for any inconvenience caused. I didn't notice jira is optional. I will make a follow-up to restore this.

**Pull request closed without merging**